plugins/dead_code_elimination/dead_code_diff.ml

```diff
@@ -0,0 +1,113 @@
+open Core_kernel
```

Comment:
Ok, we need to do it better. Currently we have the following problems:
- [ ] too complex - I want you to target to 60 lines of code or so
- [ ] ineffective - the implementation is incredibly hungry for memory and is doing too much unnecessary work
- [ ] dirty - use hashtables.

The right approach, is start from bottom to top. Define small and easy to understand functions,

e.g.,

```ocaml
let changes_in_def d1 d2 = 
   let different = phys_equal d1 d2 || Exp.(Def.rhs d1 <> Def.rhs d2) in
   if different Tid.Map.singleton (Term.tid d2) (`NewRhs (Def.rhs d2)) else Tid.Map.empty
```
Then implement `changes_in_jmps` which will return 
```
[> `NewCond of exp | `NewDest of exp] Tid.Map.t
```

and 


```ocaml
let diff_of_blk b1 b2 = 
    let (++) = Map.merge  ~f:merge_changes in
    let was = elt ~take:Blk.elts b1 and now = elts ~take:Blk.elts b2 in
    Map.fold2 was now ~init:[] ~f:(fun ~key:_ ~data diffs -> match data with
      | `Both (`Def d1, `Def d2) -> changes_in_defs d1 d2 ++ acc
      | `Both (`Jmp b1, `Jmp b2) -> changes_in_jmp d1 d2 ++ acc
      | _ -> acc)
```

where the `elts` function creates a map of a term subterms, e.g.,

```ocaml
let elts ~take t =  
  Seq.fold ~init:Tid.Map.empty (take t) ~f:(fun elts key data  -> Map.add elts ~key ~data)
```

Now use the same approach to compare two subroutines, i.e.,

```ocaml
let diff_of_subs s1 s2 = 
    let was = elts ~take:(Term.enum blk_t) s1 and now = elts ~take:(Term.enum blk_t) s2 in
    Map.fold2 ...
```

and... you guess, the same on the program level.